core/optaplanner-core-impl/src/main/java/org/optaplanner/core/api/score/buildin/bendable/BendableScore.java

```diff
@@ -301,6 +301,19 @@ public BendableScore negate() {
         return new BendableScore(-initScore, newHardScores, newSoftScores);
     }
 
+    @Override
+    public BendableScore abs() {
+        int[] newHardScores = new int[hardScores.length];
+        int[] newSoftScores = new int[softScores.length];
+        for (int i = 0; i < newHardScores.length; i++) {
+            newHardScores[i] = Math.abs(hardScores[i]);
+        }
+        for (int i = 0; i < newSoftScores.length; i++) {
+            newSoftScores[i] = Math.abs(softScores[i]);
+        }
+        return new BendableScore(Math.abs(initScore), newHardScores, newSoftScores);
```

Comment:
Init score is *never* positive. It is either negative (1 or more uninitialized entities), or zero. This is a bug. (Please also fix in other score impls.)

Reply:
Thank you for replying so fast. I am a bit confused about why this is a bug. I understand the init score is never positive. But to make sure the water level keep increasing, the increment should be positive. I didn't mean to change the score. Math.abs(initScore) is just used to make sure the increment is positive. Before I modified the GreatDelugeAcceptor, the function negate() was used there.  If the init score is never positive, then "-initScore" and "Math.abs(initScore)" should has the same result.

Reply:
The issue here is that `abs()` as implemented here can be used in any context, not just in the context of great deluge. And if someone did, they would get an init score with a positive value, which is not allowed.

However, since great deluge is a local search algorithm and it will never be allowed to run on an uninitialized solution, init score will always be zero once we get to the point when abs() would actually be called.

Therefore what I'd suggest is that abs() not touch the init score, keep what was there before. If we have the same issue in negate(), I'll look into it.

Reply:
That said, now that I look at the code we have in score implementations, it seems that indeed even multiply() will multiply the init score. That makes zero sense to me, as init score really is just a metric of how many entities are uninitialized.

I will have to talk to others about this, as I really do not understand why we do this seemingly nonsensical thing.

Reply:
No worries. All I have done is just want to make sure the great deluge algorithm works well. I indeed didn’t think about other usages of abs(). For me it is a mathematical tool for score adjustment, like power(), negate(), etc. 

Btw, I quite new on github and such a huge project. It will be nicer if you can explain more on “introduce test coverage”, then I will see what I can do on it.  I do checked this method with the employee rostering problem. It works well.

Reply:
Take a look at `HardSoftScoreTest`:
https://github.com/kiegroup/optaplanner/blob/main/core/optaplanner-core-impl/src/test/java/org/optaplanner/core/api/score/buildin/hardsoft/HardSoftScoreTest.java

There are tests for most of the methods of the `Score` interface.
We need a test added for `abs()` here, and for all the other score implementations.
Every new functionality needs to have a test, so that we can ensure we do not break it in the future.

Reply:
Thank you. I see. Tests for abs() are now added in. They are all passed for each score implementation.